backend/editor/entries.py

```diff
@@ -82,59 +84,76 @@ async def create_node(self, label, entry, main_language_code):
+            base_url = (
+                "https://raw.githubusercontent.com/openfoodfacts/openfoodfacts-server"
+                "/main/taxonomies/"
+            )
+            filename = f"{self.taxonomy_name}.txt"
+            base_url += filename
+        else:  # taxonomy is uploaded
+            filename = uploadfile.filename
 
-    async def import_from_github(self, description):
-        """
-        Helper function to import a taxonomy from GitHub
-        """
-        base_url = (
-            "https://raw.githubusercontent.com/openfoodfacts/openfoodfacts-server"
-            "/main/taxonomies/"
-        )
-        filename = self.taxonomy_name + ".txt"
-        base_url += filename
         try:
             with tempfile.TemporaryDirectory(prefix="taxonomy-") as tmpdir:
                 # File to save the downloaded taxonomy
                 filepath = f"{tmpdir}/{filename}"
+                if uploadfile is None:
+                    # Downloads and creates taxonomy file in current working directory
+                    urllib.request.urlretrieve(base_url, filepath)
+                else:
+                    with open(filepath, "wb") as f:
+                        shutil.copyfileobj(uploadfile.file, f)
+
+                with SyncTransactionCtx() as session:
+                    # Create parser object and pass current session to it
+                    parser_object = parser.Parser(session)
+                    try:
+                        # Parse taxonomy with given file name and branch name
+                        parser_object(filepath, self.branch_name, self.taxonomy_name)
+                        return True
+                    except Exception as e:
+                        raise TaxonomyParsingError() from e
+        except Exception as e:
+            raise TaxonomyImportError() from e
 
-                # Downloads and creates taxonomy file in current working directory
-                urllib.request.urlretrieve(base_url, filepath)
+    async def import_from_github(self, description, background_tasks: BackgroundTasks):
+        """
+        Helper function to import a taxonomy from GitHub
+        """
+        # Close current transaction to use the session variable in parser
+        await get_current_transaction().commit()
 
-                status = await self.parse_taxonomy(filepath)  # Parse the taxonomy
+        # Add the task to background tasks
+        background_tasks.add_task(self.parse_taxonomy)
+        print("Background task added")
 
-            async with TransactionCtx():
-                await self.create_project(description)  # Creates a "project node" in neo4j
+        # Create a new transaction context and create a "project node" in Neo4j
+        async with TransactionCtx():
+            await self.create_project(description)
 
-            return status
-        except Exception as e:
-            raise TaxonomyImportError() from e
+        return True
```

Comment:
You should do it before closing transaction and launching background task. It's more logical.

Also I don't think you need to close the transaction, as the session used  by parse_taxonomy is completely different.

Just create the project and launch the task.

```python
    await self.create_project(description)
    background_tasks.add_task(self.parse_taxonomy)
    return True
```